magi/agents/mpo/learning.py

```diff
@@ -0,0 +1,269 @@
+        target_critic_update_period: int,
+        policy_loss_fn: Optional[losses.MPO] = None,
+        counter: Optional[counting.Counter] = None,
+        logger: Optional[loggers.Logger] = None,
+    ):
+        policy_loss_fn: losses.MPO = policy_loss_fn or losses.MPO(
+            epsilon=1e-1,
+            epsilon_penalty=1e-3,
+            epsilon_mean=1e-3,
+            epsilon_stddev=1e-6,
+            init_log_temperature=1.0,
+            init_log_alpha_mean=1.0,
+            init_log_alpha_stddev=10.0,
+        )
+
+        def compute_loss(
+            policy_params: networks_lib.Params,
+            mpo_params: losses.MPOParams,
+            critic_params: networks_lib.Params,
+            target_policy_params: networks_lib.Params,
+            target_critic_params: networks_lib.Params,
+            transitions: acme_types.Transition,
+            key: jax_types.PRNGKey,
+        ):
+            # Maybe transform the observation before feeding into policy and critic.
+            # Transforming the observations this way at the start of the learning
+            # step effectively means that the policy and critic share observation
+            # network weights.
+            o_tm1 = transitions.observation
+            # This stop_gradient prevents gradients to propagate into the target
+            # observation network. In addition, since the online policy network is
+            # evaluated at o_t, this also means the policy loss does not influence
+            # the observation network training.
+            o_t = jax.lax.stop_gradient(transitions.next_observation)
+
+            # Get action distributions from policy networks.
+            online_action_distribution = policy_network.apply(policy_params, o_t)
+            target_action_distribution = policy_network.apply(target_policy_params, o_t)
+
+            # Get sampled actions to evaluate policy; of size [N, B, ...].
+            sampled_actions = target_action_distribution.sample(num_samples, seed=key)
+            tiled_o_t = utils.tile_nested(o_t, num_samples)  # [N, B, ...]
+
+            # Compute the target critic's Q-value of the sampled actions in state o_t.
+            sampled_q_t = jax.vmap(critic_network.apply, (None, 0, 0))(
+                target_critic_params,
+                tiled_o_t,
+                sampled_actions,
+            )
+
+            q_t = jnp.mean(sampled_q_t, axis=0)  # [B]
+
+            # Compute online critic value of a_tm1 in state o_tm1.
+            q_tm1 = critic_network.apply(
+                critic_params, o_tm1, transitions.action
+            )  # [B]
+            # q_tm1 = tf.squeeze(q_tm1, axis=-1)  # [B]; necessary for trfl.td_learning.
+
+            # Critic loss.
+            batch_td_learning = jax.vmap(rlax.td_learning)
+            td_error = batch_td_learning(
+                q_tm1, transitions.reward, discount * transitions.discount, q_t
+            )
+            critic_loss = jnp.mean(jnp.square(td_error))
```

Comment:
That's interesting, I have seen some sort of filtering when working with sequences but usually, those are done in the policy. 

Do you mean something like?
```
critic_loss = jnp.mean(jnp.square(td_error) * transitions.discount)
```
This would filter out the the transitions into a terminal state.